### PR TITLE
Add the possibility to create the metric reporter topic at launch

### PR DESCRIPTION
**Pull request by BDeus**
I use this code to create topic "__CruiseControlMetrics" in cruisecontrolmetricsreporter.
It can deploy the topic with parameterized config in case of a non auto creation topic Cluster.

### CODE REVIEW

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java [hunk 1]
```diff
@@ -112,11 +119,46 @@ public void configure(Map<String, ?> configs) {
 
     _cruiseControlMetricsTopic = reporterConfig.getString(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_CONFIG);
     _reportingIntervalMs = reporterConfig.getLong(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_REPORTING_INTERVAL_MS_CONFIG);
+
+    if(reporterConfig.getBoolean(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_AUTO_CREATE_CONFIG)) {
+      _newTopic = createNewTopicFromReporterConfig(reporterConfig);
+      final Properties adminProperties = new Properties();
+      adminProperties.put(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG, producerProps.get(CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG));
+      adminProperties.put(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG, producerProps.get(CommonClientConfigs.SECURITY_PROTOCOL_CONFIG));
```
Comment (efeg) [inline cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java]: If the security protocol is `SSL`, I suspect that the `AdminClient` properties set here will be missing a number of required configs. (please see a similar patch in `migrate_to_kafka_2_0` branch of Cruise Control: https://github.com/linkedin/cruise-control/commit/2d7b519d806a7a2228a2b19fdedf6a457e6053d8)

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java [hunk 2]
```diff
@@ -112,11 +119,46 @@ public void configure(Map<String, ?> configs) {
...
+  }
+
+  private static NewTopic createNewTopicFromReporterConfig(CruiseControlMetricsReporterConfig reporterConfig) {
+    String cruiseControlMetricsTopic = reporterConfig.getString(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_CONFIG);
+    Integer cruiseControlMetricsTopicNumPartition = reporterConfig.getInt(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_NUM_PARTITIONS_CONFIG);
+    Short cruiseControlMetricsTopicReplicaFactor = reporterConfig.getShort(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_REPLICATION_FACTOR_CONFIG);
+    return new NewTopic(cruiseControlMetricsTopic, cruiseControlMetricsTopicNumPartition, cruiseControlMetricsTopicReplicaFactor);
+  }
+
+  private static void createCruiseControlMetricsTopic(AdminClient adminClient, NewTopic newTopic) {
```
Comment (efeg) [inline cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java]: Would it possible to:
1. Remove `static` keyword,
2. Do not take any argument, and
3. User `_adminClient` and `_newTopic`?

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java [hunk 3]
```diff
@@ -112,11 +119,46 @@ public void configure(Map<String, ?> configs) {
...
+    try {
+      final CreateTopicsResult createTopicsResult = adminClient.createTopics(Collections.singletonList(newTopic));
+      createTopicsResult.values().get(newTopic.name()).get();
+      LOG.info("Cruise Control metrics topic created");
```
Comment (efeg) [inline cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java]: Nit: Can we also log the name of the topic?

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporterConfig.java [hunk 4]
```diff
@@ -13,20 +13,29 @@
...
+  public static final String CRUISE_CONTROL_METRICS_TOPIC_REPLICATION_FACTOR_CONFIG = "cruise.control.metrics.topic.replication.factor";
+  private static final String CRUISE_CONTROL_METRICS_TOPIC_REPLICATION_FACTOR_DOC = "The replication factor of Cruise Control metrics topic";
   public static final String CRUISE_CONTROL_METRICS_REPORTING_INTERVAL_MS_CONFIG = PREFIX + "metrics.reporting.interval.ms";
   private static final String CRUISE_CONTROL_METRICS_REPORTING_INTERVAL_MS_DOC = "The interval in milliseconds the "
       + "metrics reporter should report the metrics.";
   // Default values
   public static final String DEFAULT_CRUISE_CONTROL_METRICS_TOPIC = "__CruiseControlMetrics";
+  public static final Integer DEFAULT_CRUISE_CONTROL_METRICS_TOPIC_NUM_PARTITIONS = 1;
+  public static final boolean DEFAULT_CRUISE_CONTROL_METRICS_TOPIC_AUTO_CREATE = false;
+  public static final Short DEFAULT_CRUISE_CONTROL_METRICS_TOPIC_REPLICATION_FACTOR = 1;
```
[the remark this answers is not in the document]
Comment (BDeus) [reply cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporterConfig.java]: Yes it safer to let the default value to something like -1 and let the user to explicitly set the parameters

File: cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java (hunk 1 above)
Comment (BDeus) [reply cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java]: The producer already not set all this SSL properties. Does it currently work with SSL?

Comment (efeg) [reply cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java]: -- I agree that it is not clear from the code -- I'll add pointers/comments to make it more clear. 
The [documentation on configurations](https://github.com/linkedin/cruise-control/wiki/Configurations) indicate that `CruiseControlMetricsReporter takes all the configurations for vanilla KafkaProducer with a prefix of "cruise.control.metrics.reporter."`.

**Comment by BDeus** [comment]
I have address the few nits. You can patch when you want.

**Pull request merged**